File: internal/api/jdwp/modulereference.go
package jdwp

import (
	"fmt"

	"cli-debugger/pkg/errors"
)

// ModuleReference Command Set Implementation
// ModuleReference command constants are already defined in protocol.go

// Name Get module name
func (c *Client) ModuleName(moduleID string) (string, error) {
	data := encodeID(moduleID, c.idsizes.ObjectIDSize)

	packet := createCommandPacketWithData(18, 1, data)
	if err := c.sendPacket(packet); err != nil {
		return "", errors.WrapCommandError(err, errors.ErrCommandFailed, "Failed to get module name")
	}

	reply, err := c.readReply()
	if err != nil {
		return "", errors.WrapCommandError(err, errors.ErrCommandFailed, "Failed to get module name")
	}

	if reply.ErrorCode != 0 {
		return "", errors.NewProtocolError(errors.ErrProtocolError,
			fmt.Sprintf("Get module name failed: %s", reply.Message))
	}

	reader := newPacketReader(reply.Data)
	moduleName, err := reader.readString()
	if err != nil {
		return "", errors.WrapProtocolError(err, errors.ErrInvalidResponse, "Failed to read module name")
	}

	return moduleName, nil
}

// ClassLoader Get module class loader
func (c *Client) ModuleClassLoader(moduleID string) (string, error) {
	data := encodeID(moduleID, c.idsizes.ObjectIDSize)

	packet := createCommandPacketWithData(18, 2, data)
	if err := c.sendPacket(packet); err != nil {
		return "", errors.WrapCommandError(err, errors.ErrCommandFailed, "Failed to get module class loader")
	}

	reply, err := c.readReply()
	if err != nil {
		return "", errors.WrapCommandError(err, errors.ErrCommandFailed, "Failed to get module class loader")
	}

	if reply.ErrorCode != 0 {
		return "", errors.NewProtocolError(errors.ErrProtocolError,
			fmt.Sprintf("Get module class loader failed: %s", reply.Message))
	}

	reader := newPacketReader(reply.Data)
	classLoaderID := reader.readID(c.idsizes.ObjectIDSize)

	return classLoaderID, nil
}

File: src/protocol/jdwp/module-reference.ts
/**
 * ModuleReference Command Set Implementation
 */

import {
  CommandSet,
  ModuleReferenceCommand,
  type IDSizes,
} from "./protocol.js";
import {
  createCommandPacketWithData,
  encodeID,
} from "./codec.js";
import { PacketReader } from "./reader.js";
import { APIError, ErrorType, ErrorCodes } from "../errors.js";

/**
 * JDWP Client interface for command execution
 */
export interface JDWPCommandExecutor {
  sendPacket(packet: Buffer): Promise<void>;
  readReply(): Promise<{ errorCode: number; message: string; data: Buffer }>;
  idSizes: IDSizes;
}

/**
 * Get module name
 */
export async function getModuleName(
  executor: JDWPCommandExecutor,
  moduleID: string,
): Promise<string> {
  const data = encodeID(moduleID, executor.idSizes.objectIDSize);
  const packet = createCommandPacketWithData(
    CommandSet.ModuleReference,
    ModuleReferenceCommand.Name,
    data,
  );
  await executor.sendPacket(packet);

  const reply = await executor.readReply();
  if (reply.errorCode !== 0) {
    throw new APIError(
      ErrorType.ProtocolError,
      ErrorCodes.ProtocolError,
      `Get module name failed: ${reply.message}`,
    );
  }

  const reader = new PacketReader(reply.data);
  return reader.readString();
}

/**
 * Get module class loader
 */
export async function getModuleClassLoader(
  executor: JDWPCommandExecutor,
  moduleID: string,
): Promise<string> {
  const data = encodeID(moduleID, executor.idSizes.objectIDSize);
  const packet = createCommandPacketWithData(
    CommandSet.ModuleReference,
    ModuleReferenceCommand.ClassLoader,
    data,
  );
  await executor.sendPacket(packet);

  const reply = await executor.readReply();
  if (reply.errorCode !== 0) {
    throw new APIError(
      ErrorType.ProtocolError,
      ErrorCodes.ProtocolError,
      `Get module class loader failed: ${reply.message}`,
    );
  }

  const reader = new PacketReader(reply.data);
  return reader.readID(executor.idSizes.objectIDSize);
}
